src/types.jl

```diff
@@ -535,8 +554,8 @@ end
 
 unflatten(t) = t
 
-function TermInterface.maketerm(::Type{<:BasicSymbolic}, head, args, type, metadata)
-    basicsymbolic(head, args, type, metadata)
+function TermInterface.maketerm(T::Type{<:BasicSymbolic}, head, args, metadata)
+    basicsymbolic(head, args, symtype(T), metadata)
```

Comment:
This is not always true. Consider
```julia
ex = x + y
maketerm(typeof(ex), ==, arguments(ex), metadata(ex))
```
The current implementation would derive `x == y` to be `Real` while it should be `Bool`.